Unify EdgeActions with CDN 2025-09-01-preview and migrate to TypeSpec

tundwed opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![spec_pr_review_workflow_diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5bb5e7ce-8aff-4dbb-a3f8-0d9b68fef5b1)

## Purpose of this PR

We're adding a new API version: We copied the EdgeActions specification from 2024-07-22-preview to 2025-09-01-preview.

Version alignment purpose: The new API version enables unified SDK generation for both CDN and EdgeActions services for edge actions public preview

  - [ ] New resource provider.
  - [x ] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Convert existing [OpenAPI spec to TypeSpec spec](https://aka.ms/typespec/conversion) (do not combine this with implementing changes for a new API version).
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x ] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x ] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.
- [ x] A [release plan](https://aka.ms/azsdkdocs/release-plans) has been created. If not, please create one as it will help guide you through the REST API and SDK creation process. 

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories)
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.
- For guidance on SDK breaking change review, refer to https://aka.ms/ci-fix.

mattmccreesh reviewed (commented):
This references a manifest version not deployed to ARM manifest yet. Will this cause validation issues? Will SDK be generated in broken state? Lets confirm this. I'm okay with releasing in parallel to meet public preview timelines, but lets get process right after public preview.

mikeharder reviewed (dismissed):
must convert to typespec

tundwed commented:
> must convert to typespec
@mikekistler converted to typespec

ramoka178 commented on specification/cdn/resource-manager/Microsoft.Cdn/EdgeActions/readme.md:
```diff
@@ -0,0 +1,94 @@
+  - code: ArmResourcePropertiesBag
+    reason: This is the API design and therefore exempted
+  - code: ConsistentPatchProperties
+    reason: This is the API design and therefore exempted
```
please include where claus instead of suppressing on a whole.

tundwed replied:
added where clauses

Commit:
Add EdgeActions 2025-09-01-preview baseline from 2024-07-22-preview

Copy EdgeActions API specification from 2024-07-22-preview to new EdgeActions directory structure as baseline for 2025-09-01-preview.

Commit:
Migrate EdgeActions to TypeSpec with API changes for 2025-09-01-preview

- Convert EdgeActions from JSON swagger to TypeSpec
- Add TypeSpec configuration for all SDK languages (Python, Java, Go, TypeScript, C#)
- Configure separate Azure.ResourceManager.EdgeActions package for .NET SDK
- Add @clientName decorator for EdgeActionAttachmentResponse -> EdgeActionAttachmentResult (C# naming convention)
- Add file-scoped suppression for ProvisioningStateMustBeReadOnly
- Add new operations: SwapDefault, DeployVersionCode, GetVersionCode
- Update all examples to 2025-09-01-preview
- Generate OpenAPI via TypeSpec compiler with automatic x-ms-examples

weidongxu-microsoft commented on specification/cdn/resource-manager/Microsoft.Cdn/EdgeActions/tspconfig.yaml:
```diff
@@ -0,0 +1,58 @@
+    service-name: "EdgeActions"
+    premium: true
+    service-dir: "sdk/edgeactions"
+    enable-sync-stack: false
```
remove `premium` and `enable-sync-stack` option

razvanbadea-msft replied:
premium is still present

Commit:
fix java sdk validation

razvanbadea-msft commented on specification/cdn/resource-manager/Microsoft.Cdn/EdgeActions/readme.md:
```diff
@@ -0,0 +1,89 @@
+    where: $.paths["/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Cdn/edgeActions/{edgeActionName}/versions/{version}/swapDefault"].post
+    reason: Preexisting LRO pattern (200,202) retained for backward compatibility with 2024-07-22-preview.
+  # ProvisioningState properties are all marked readOnly in the emitted spec. File-scoped suppression to cover all occurrences.
+  - code: ProvisioningStateMustBeReadOnly
```
it does not sound right, where are the cases when the errors occurs?

tundwed replied:
@razvanbadea-msft  When i remove the suppression, these are errors I get

<img width="3024" height="1716" alt="image" src="https://github.com/user-attachments/assets/5b77c27e-6c39-498d-9254-2242b6220552" />

<img width="2946" height="1653" alt="image" src="https://github.com/user-attachments/assets/71e01e27-bc0b-4542-a241-91be48e28d24" />


If you look at the line numbers, it points to, like line 777, these are responses that point to schemas whose models have provisioning state in their properties. Provisioning state in those properties is set to readonly. but it does not seem to pick up on that. I think it would go away if we were to make the enum itself readonly, but i do not know of a way to do that in typespec, only on properties. 

<img width="2061" height="426" alt="image" src="https://github.com/user-attachments/assets/feb7d8e8-8349-4a25-a3c9-33b139b550fb" />

For now, I have put back the suppression. Let me know if there is a different way to do this

tundwed replied:
Removed suppression and used this recommended workaround

https://github.com/Azure/azure-rest-api-specs/wiki/Swagger-LintDiff#provisioningstatemustbereadonly

razvanbadea-msft commented on specification/cdn/resource-manager/Microsoft.Cdn/EdgeActions/preview/2025-09-01-preview/openapi.json:
```diff
@@ -0,0 +1,2067 @@
+      "description": "Concrete tracked resource types can be created by aliasing this type using a specific property type.",
+      "properties": {
+        "properties": {
+          "$ref": "#/definitions/EdgeActionProperties",
```
in base version the x-ms-mutability was set like now it is not

"x-ms-client-flatten": true,
          "x-ms-mutability": [
            "read",
            "create"
          ]

razvanbadea-msft commented on specification/cdn/resource-manager/Microsoft.Cdn/EdgeActions/preview/2025-09-01-preview/openapi.json:
```diff
@@ -0,0 +1,2067 @@
+      "properties": {
+        "properties": {
+          "$ref": "#/definitions/EdgeActionExecutionFilterProperties",
+          "description": "The resource-specific properties for this resource."
```
in base version

"x-ms-client-flatten": true,
          "x-ms-mutability": [
            "read",
            "create"
          ]

razvanbadea-msft commented on specification/cdn/resource-manager/Microsoft.Cdn/EdgeActions/preview/2025-09-01-preview/openapi.json:
```diff
@@ -0,0 +1,2067 @@
+      "properties": {
+        "properties": {
+          "$ref": "#/definitions/EdgeActionVersionProperties",
+          "description": "The resource-specific properties for this resource."
```
in base version

"x-ms-client-flatten": true,
          "x-ms-mutability": [
            "read",
            "create"
          ]

razvanbadea-msft commented on specification/cdn/resource-manager/Microsoft.Cdn/EdgeActions/preview/2025-09-01-preview/openapi.json:
```diff
@@ -0,0 +1,2067 @@
+        "Provisioning",
+        "Upgrading"
+      ],
+      "x-ms-enum": {
```
ProvisioningState does not have  readonly set to true

tundwed replied:
@razvanbadea-msft 

The ProvisioningState definition (lines 1970-1977) is a type definition and doesn't have readonly

All properties that use ProvisioningState already have "readOnly": true set (e.g., line 1738 in EdgeActionProperties, line 1625 in EdgeActionExecutionFilterProperties, line 1905 in EdgeActionVersionProperties)

I believe the @visibility(Lifecycle.Read) decorator in typespec only works on properties (model fields) and not on type definitions. I could be wrong. Let me know what you think

tundwed replied:
Removed suppression and used this recommended workaround

https://github.com/Azure/azure-rest-api-specs/wiki/Swagger-LintDiff#provisioningstatemustbereadonly

tundwed commented:
@microsoft-github-policy-service agree company="Microsoft"

Commit:
fix(edgeactions): address PR review comments for 2025-09-01-preview

- Add x-ms-client-flatten and x-ms-mutability to properties field on EdgeAction, EdgeActionVersion, and EdgeActionExecutionFilter resource models for backwards compatibility with 2024-07-22-preview baseline
- Change EdgeActionAttachmentResponse.edgeActionId from readOnly to x-ms-mutability [read, update, create] to match baseline behavior
- Replace 'ARM' with 'Azure' in resource identifier descriptions (AttachedResourceId and EdgeActionVersionId)
- Consolidate multiple PostResponseCodes suppressions into single entry with array of where clauses in readme.md
- Add OpenAPI extension decorators using @@extension augments in TypeSpec with appropriate suppressions for no-openapi warnings

Commit:
style(edgeactions): apply tsp format to TypeSpec files

- Reformat @@extension calls to multi-line style for better readability
- Reorder #suppress directive placement per formatter rules

tundwed commented on specification/cdn/resource-manager/Microsoft.Cdn/EdgeActions/edgeactionsbasemodels.tsp:
```diff
@@ -0,0 +1,240 @@
+    }
+  ]>;
+
+@doc("ARM resource identifier for edge action versions")
```
done

Commit:
remove unnecessary ProvisioningStateMustBeReadOnly suppression

All provisioningState properties are already correctly marked as readOnly in the generated OpenAPI spec. The suppression is not needed as the spec complies with ARM validation rules.

Commit:
Revert "remove unnecessary ProvisioningStateMustBeReadOnly suppression"

This reverts commit a5988e2e2c8188e0836b5d366bb19469535ca772.

Commit:
extend ResourceProvisioningState for standard ARM compliance

Change ProvisioningState to extend ResourceProvisioningState which provides the three required states (Succeeded, Failed, Canceled) with standard ARM descriptions, and add custom states (Provisioning, Upgrading) specific to edge actions lifecycle.

Commit:
remove ProvisioningStateMustBeReadOnly suppression

Suppression no longer needed after extending ResourceProvisioningState which properly handles ARM validation requirements.

Commit:
add use-read-only-status-schema to resolve ProvisioningStateMustBeReadOnly validation

Enable use-read-only-status-schema: true in tspconfig.yaml to work around known limitation in azure-openapi-validator and oav with \ siblings. This adds readOnly: true directly to status schema definitions, resolving ProvisioningStateMustBeReadOnly validation errors.

References:
- https://github.com/Azure/azure-openapi-validator/issues/637
- https://github.com/Azure/oav/issues/848

Ptnan7 reviewed (approved):
Approve after validating in generating pwsh and cli
